### PR TITLE
Structured network compute plus reservoirs

### PR DESCRIPTION
**Pull request by jameshalgren**
This pull request introduces the capability to perform reservoir simulation in the v02 framework. The new functions `compute_network_structured_obj` is currently the only version of the function which permits this simulation and at present, it only allows works with the by-network and serial parallel methods

The Reservoir code includes unit tests, with instructions in src/fortran_routing/Reservoir_Binding/kernels/readme.md

The following commandline tests execute successfully: 
- [x] `python compute_nhd_routing_SingleSeg_v02.py --test pocono1 -t -v --debuglevel 1`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/CustomInput.yaml`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -t -n Mainstems_CONUS -v --debuglevel 1 --parallel by-network --nts 288 --cpu-pool 6`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -t -n Mainstems_CONUS -v --debuglevel 1 --parallel by-subnetwork-jit-clustered --subnet-size 100 --nts 288 --cpu-pool 6`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/Florence_Benchmark.yaml`

The CustomInput.yaml example tests the newly added reservoir capability. Please note the caveats from the comments in that file: 
```
  break_network_at_waterbodies: true # replace waterbodies in the route-link dataset with segments representing the reservoir and calculate to divide the computation (leave blank for
  # WARNING: `break_network_at_waterbodies: true` will only work if compute_method is set to "V02-structured-obj" and parallel_compute_method is unset (serial execution) or set to "by-network".
  compute_method: V02-structured-obj  # OPTIONS: "V02-caching", "V02-structured-obj", "V02-structured"
  return_courant: false  # WARNING: true will only work with compute_method "V02-caching", therefore not currently compatible with simulation for waterbodies.

```

### CODE REVIEW

**Comment by jameshalgren** [comment]
Executing at test case shows that the current cache-based method produces results that are [almost] identical to the output from the original WRF-Hydro model output. 
```
(mesh) [james.halgren@nwcal-ihcb-bm01 python_routing_v02]$ python compute_nhd_routing_SingleSeg_v02.py --test pocono1 -v --compute-method V02-caching -t
running test case for Pocono_TEST1 domain - NO RESERVOIRS
creating supernetwork connections set
supernetwork connections set complete
... in 0.027332067489624023 seconds.
organizing connections into reaches ...
reach organization complete
... in 0.0007894039154052734 seconds.
setting channel initial states ...
channel initial states complete
... in 0.01580214500427246 seconds.
creating qlateral array ...
qlateral array complete
... in 0.3329293727874756 seconds.
executing routing computation ...
ordered reach computation complete
... in 0.55668044090271 seconds.
conducting parity check, comparing WRF Hydro results against t-route results
                     flow, wrf (cms)  flow, t-route (cms)
2018-01-01 02:00:00         0.873438             0.873438
2018-01-01 03:00:00         0.873347             0.873347
2018-01-01 04:00:00         0.873250             0.873250
2018-01-01 05:00:00         0.873148             0.873148
2018-01-01 06:00:00         0.873040             0.873040
2018-01-01 07:00:00         0.872928             0.872928
2018-01-01 08:00:00         0.872810             0.872810
2018-01-01 09:00:00         0.872687             0.872687
2018-01-01 10:00:00         0.872559             0.872558
2018-01-01 11:00:00         0.872425             0.872425
2018-01-01 12:00:00         0.872288             0.872288
2018-01-01 13:00:00         0.872145             0.872145
2018-01-01 14:00:00         0.871998             0.871998
2018-01-01 15:00:00         0.871845             0.871845
2018-01-01 16:00:00         0.871689             0.871688
2018-01-01 17:00:00         0.871527             0.871527
2018-01-01 18:00:00         0.871361             0.871361
2018-01-01 19:00:00         0.871191             0.871191
2018-01-01 20:00:00         0.871015             0.871016
2018-01-01 21:00:00         0.870836             0.870837
2018-01-01 22:00:00         0.870653             0.870653
2018-01-01 23:00:00         0.870465             0.870465
2018-01-02 00:00:00         0.870273             0.870273
2018-01-02 01:00:00         0.870077             0.870077
```
However, the structured function does not exhibit the same level of parity. 
```
(mesh) [james.halgren@nwcal-ihcb-bm01 python_routing_v02]$ python compute_nhd_routing_SingleSeg_v02.py --test pocono1 -v --compute-method V02-structured -t
running test case for Pocono_TEST1 domain - NO RESERVOIRS
creating supernetwork connections set
supernetwork connections set complete
... in 0.027153968811035156 seconds.
organizing connections into reaches ...
reach organization complete
... in 0.0007994174957275391 seconds.
setting channel initial states ...
channel initial states complete
... in 0.015575408935546875 seconds.
creating qlateral array ...
qlateral array complete
... in 0.33203887939453125 seconds.
executing routing computation ...
ordered reach computation complete
... in 0.5351865291595459 seconds.
conducting parity check, comparing WRF Hydro results against t-route results
                     flow, wrf (cms)  flow, t-route (cms)
2018-01-01 02:00:00         0.873438             0.873262
2018-01-01 03:00:00         0.873347             0.872736
2018-01-01 04:00:00         0.873250             0.872574
2018-01-01 05:00:00         0.873148             0.872427
2018-01-01 06:00:00         0.873040             0.872275
2018-01-01 07:00:00         0.872928             0.872119
2018-01-01 08:00:00         0.872810             0.871958
2018-01-01 09:00:00         0.872687             0.871792
2018-01-01 10:00:00         0.872559             0.871622
2018-01-01 11:00:00         0.872425             0.871448
2018-01-01 12:00:00         0.872288             0.871269
2018-01-01 13:00:00         0.872145             0.871085
2018-01-01 14:00:00         0.871998             0.870898
2018-01-01 15:00:00         0.871845             0.870706
2018-01-01 16:00:00         0.871689             0.870510
2018-01-01 17:00:00         0.871527             0.870310
2018-01-01 18:00:00         0.871361             0.870106
2018-01-01 19:00:00         0.871191             0.869898
2018-01-01 20:00:00         0.871015             0.869686
2018-01-01 21:00:00         0.870836             0.869469
2018-01-01 22:00:00         0.870653             0.869250
2018-01-01 23:00:00         0.870465             0.869026
2018-01-02 00:00:00         0.870273             0.868798
2018-01-02 01:00:00         0.870077             0.868567
(mesh) [james.halgren@nwcal-ihcb-bm01 python_routing_v02]$

```

**Comment by jameshalgren** [comment]
With last two commits, solution is much closer to full parity. Still may be some inconsistency in the handling of initial states. Checking into that...

**Comment by jameshalgren** [comment]
@jdmattern-noaa 
There are, as we discussed, a number of relevant Git Items to bring to the table to pull the reservoirs into this. Our first step might be to make the reservoirs work with the V02-structured-obj CLI option -- this is the simplest of the compute_network functions, even though it is the least performant of the v02 options. 

__Relevant links:__
* Code to separate Waterbodies -- see nhd_io.py for reading; nhd_network.py, last four functions for execution and network adjustment for reservoirs.
https://github.com/groutr/t-route/blob/waterbodies/src/python_framework_v02/nhd_network.py
* Latest structured functions to facilitate distinguishing Reservoir and Routing Reaches
https://github.com/NOAA-OWP/t-route/pull/224/files
    * (original PR for the structured functions: https://github.com/NOAA-OWP/t-route/pull/165/files#)
* Example from V01 of inserting Reservoir Data Handling: 
https://github.com/groutr/t-route/blob/a521ea3ac6e60960f01a1daa024e33a1a7120366/src/python_routing_v01/compute_nhd_routing_SingleSeg.py#L1346

Key notes: 
Line 434 of the current nhd_network_utilities_v02.py contains the function: def organize_independent_networks(connections):

This is parallel to what happens in lines 209-215 in the waterbodies PR #76 from Ryan (link above). We need to add the same function as a new function or as logic with in the organize_independent_networks.

**Comment by hellkite500** [comment]
@groutr Maybe you can help elaborate on this a bit.  Trying  to connect a reach to a waterbody, looking at these structures:

```python
# STEP 1: Build basic network connections graph
    connections, wbodies, param_df = nnu.build_connections(supernetwork_parameters, dt)
    connections = nnu.replace_waterbodies_connections(connections, wbodies)
    if verbose:
        print("supernetwork connections set complete")
    if showtiming:
        print("... in %s seconds." % (time.time() - start_time))

    # STEP 2: Identify Independent Networks and Reaches by Network
    if showtiming:
        start_time = time.time()
    if verbose:
        print("organizing connections into reaches ...")

    independent_networks, reaches_bytw, rconn = nnu.organize_independent_networks(
        connections
    )

    for k,v  in wbodies.items():
        #print(k, v)
        if k in  reaches_bytw.keys():
            print("found wbodies key  in  reaches keys")
        if k in reaches_bytw.values():
            print("found wbodies key  in  reaches values")
        if v in  reaches_bytw.keys():
            print("found wbodies value  in  reaches keys")
        if v in reaches_bytw.values():
            print("found wbodies value  in  reaches values")
        os._exit(1)
```
If I understand correctly, reaches_bytw is a dictionary with keys of the reach's  tailwater, and values are  the list of reaches (list of list of segments, `[ [reach1.seg1, reach1.seg2], [reach2.seg1, reach2.seg2] ... ]`
I want to find  which of those reaches are waterbodies.   The intersection you described I thought was between the `wbodies` which comes from a call to
```python
wbodies = nhd_network.extract_waterbodies(
        param_df, cols["waterbody"], supernetwork_parameters["waterbody_null_code"]
    )
```
inside `build_connections`, and the network reaches.  But  there is no intersection in either they keys or values of `reaches_bytw`.   Is this not the correct intersection?  Does it  need  to  be the `independent_networks`?  I checked that too, didn't find any mapping.  For completeness, I found ONE of multiple waterbodies keys in `rconn.values` but I don't really understand how that is  useful at the moment.

**Comment by jameshalgren** [comment]
The keys of reaches_bytw are the tailwaters of independent networks. 
Try this:
```
In : for _, reach_list in reaches_bytw.items():
...:     for reach in reach_list:
...:         print(reach[-1] in wbodies)
```

**Comment by hellkite500** [comment]
Thanks for that, I missed a list level in my first look it seems. Will take a look at this mapping and see where to start hacking `reaches_bytw`

**Comment by hellkite500** [comment]
Also, why is this reach a waterbody but has two segments?
`[4185665, 4186287], True`

**Comment by jameshalgren** [comment]
> Also, why is this reach a waterbody but has two segments?
> `[4185665, 4186287], True`

We need to use the [`split_at_waterbodies_and_junctions`](https://github.com/NOAA-OWP/t-route/blob/7a923182987515eeaf3a0e14baf5a678c54807a0/src/python_framework_v02/troute/nhd_network.py#L173)

in the [`organize_independent_networks` function](https://github.com/NOAA-OWP/t-route/blob/7a923182987515eeaf3a0e14baf5a678c54807a0/src/python_framework_v02/troute/nhd_network_utilities_v02.py#L440)

And then we need to coalesce the waterbodies into a single representative segment with [`replace_waterbodies_connections`](https://github.com/NOAA-OWP/t-route/blob/7a923182987515eeaf3a0e14baf5a678c54807a0/src/python_framework_v02/troute/nhd_network.py#L394)

**Comment by groutr** [comment]
@hellkite500 the intersection I was describing was between the wbodies set and the graph returned from `replace_waterbodies_connections`.

**Comment by groutr** [comment]
This needs to be clarified as it seems there is some misunderstanding what `replace_waterbody_connections` actually does.

`connections` is a graph (dict[Int, List[Int]])
`waterbodies` is a dictionary with nodes as keys and _waterbody codes_ as values.

```python
def replace_waterbodies_connections(connections, waterbodies):
    """
    Use a single node to represent waterbodies. The node id is the
    waterbody id.
    This returns a new copy of connections with transformation
    """
    new_conn = {}
    waterbody_nets = separate_waterbodies(connections, waterbodies) # this doesn't seem to be used....

    for n in connections:  # Loop thru all the nodes in the graph
        if n in waterbodies:  # if our current node belongs to a waterbody
            wbody_code = waterbodies[n]  # get the waterbody code
            if wbody_code in new_conn:  # Have we already seen/processed this waterbody?
                continue

            # get all nodes that belong to the waterbody
            wbody_nodes = [k for k, v in waterbodies.items() if v == wbody_code]
            # Get all nodes immediately downstream of waterbody
            outgoing = reservoir_shore(connections, wbody_nodes)
            # add the waterbody code to the network and link it to the downstream nodes
            new_conn[wbody_code] = outgoing

        elif reservoir_boundary(connections, waterbodies, n):
            # one of the children of n is a member of a waterbody
            # replace that child with waterbody code.
            new_conn[n] = []

            for child in connections[n]:
                if child in waterbodies:
                    new_conn[n].append(waterbodies[child])
                else:
                    new_conn[n].append(child)
        else:
            # copy to new network unchanged
            new_conn[n] = connections[n]
    return new_conn
```
Here is an example:
network = `{1: [2], 2: [4], 3: [4], 4: [6, 5], 5: [7], 7: [8], 6: [], 8: []}` and waterbodies = `{2: 99, 4: 99, 5: 99}`

new_conn = `{1: [99], 3: [99], 99: [6, 7], 7: [8], 6: [], 8: []}`
Note that we have replaced all references to nodes 2, 4, and 5 with the waterbody code of 99.

To get the waterbodies, we can use the waterbody code (from the values of the waterbodies mapping)
```
wbodies = new_conn.keys() & waterbodies.values()
```

**Comment by groutr** [comment]
@jameshalgren your order of operations is reversed. You need to coalesce the waterbody nodes first, then use `split_at_waterbodies_and_junctions`. The splitting function will not work correctly if you haven't coalesced the waterbodies first.

**Comment by jameshalgren** [comment]
@groutr, I see that and you are right -- Like this, then: 
First, the changes in `/src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py`
```
+    connections = nhd_network.replace_waterbodies_connections(connections, wbodies)
     if verbose:
         print("supernetwork connections set complete")
     if showtiming:
@@ -782,8 +783,10 @@ def main():
         print("organizing connections into reaches ...")

     independent_networks, reaches_bytw, rconn = nnu.organize_independent_networks(
-        connections
+        connections,
+        wbodies,
     )
```
... and in `/src/python_framework_v02/troute/nhd_network_utilities_v02.py`
```
-def organize_independent_networks(connections):
+def organize_independent_networks(connections, wbodies=None):

     rconn = nhd_network.reverse_network(connections)
     independent_networks = nhd_network.reachable_network(rconn)
     reaches_bytw = {}
     for tw, net in independent_networks.items():
-        path_func = partial(nhd_network.split_at_junction, net)
+        if wbodies:
+            path_func = partial(nhd_network.split_at_waterbodies_and_junctions, wbod
+        else:
+            path_func = partial(nhd_network.split_at_junction, net)
         reaches_bytw[tw] = nhd_network.dfs_decomposition(net, path_func)
```
Yielding, now as expected:
```
In : for _, reach_list in reaches_bytw.items():
...:     for reach in reach_list:
...:         for r in reach:
...:             print(f"{reach}, {r in wbodies.values()}")
...:
[4185517, 4185587, 4186251], False
[4185517, 4185587, 4186251], False
[4185517, 4185587, 4186251], False
[4185527, 4185583, 4186253], False
[4185527, 4185583, 4186253], False
[4185527, 4185583, 4186253], False
[4186255, 4185607, 4186257, 4185661], False
[4186255, 4185607, 4186257, 4185661], False
[4186255, 4185607, 4186257, 4185661], False
[4186255, 4185607, 4186257, 4185661], False
[4186499, 4185663], False
[4186499, 4185663], False
[4185665], False
[4185613, 4185667], False
[4185613, 4185667], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185649, 932040253, 932040251, 932040249, 4186503, 4185693], False
[4185089], True
[4153018], False
[4153786], False
In : wbodies
{4186287: 4185089, 4186297: 4185089, 4186305: 4185089, 4186307: 4185089, 4186309: 4185089}
```

**Comment by jameshalgren** [comment]
rebased to latest upstream/master.

File: src/python_framework_v02/troute/nhd_network_utilities_v02.py [hunk 1]
```diff
@@ -416,28 +416,50 @@ def build_connections(supernetwork_parameters, dt):
...
+
+
+def organize_independent_networks(connections, wbodies=None):
 
     rconn = nhd_network.reverse_network(connections)
     independent_networks = nhd_network.reachable_network(rconn)
     reaches_bytw = {}
     for tw, net in independent_networks.items():
-        path_func = partial(nhd_network.split_at_junction, net)
+        if wbodies:
```
Comment (groutr) [inline src/python_framework_v02/troute/nhd_network_utilities_v02.py:456]: We could probably pull the partial function definitions out of the loop. Pass the entire network as an argument.

File: src/python_routing_v02/fast_reach/mc_reach.pyx [hunk 2]
```diff
@@ -783,7 +783,7 @@ cpdef object compute_network_structured_obj(
               buf_view[i][8] = segment.cs
               buf_view[i][9] = segment.s0
               buf_view[i][10] = flowveldepth[segment.id, timestep-1, 0]
-              buf_view[i][11] = flowveldepth[segment.id, timestep-1, 1]
+              buf_view[i][11] = 0.0 #flowveldepth[segment.id, timestep-1, 1]
```
Comment (groutr) [inline src/python_routing_v02/fast_reach/mc_reach.pyx:786]: Try using `buf_view[i, j]` for all of these memoryview accesses.

Comment (jameshalgren) [reply src/python_routing_v02/fast_reach/mc_reach.pyx:863]: Fixed...

**Comment by jameshalgren** [comment]
__An update on progress:__
ping: @groutr @hellkite500 @jdmattern-noaa 
I have merged the work from @jdmattern-noaa and the test now calculates successfully for the three micro reservoirs in the Pocono test dataset in this version of the repository. 
```
python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/CustomInput.yaml
```

We have at least the following outstanding items -- some of these could be managed in a separate pull request, but I'll list them here:
- [ ] Add the waterbody preprocessing to the `by-subnetwork-jit` and `by-subnetwork-jit-clustered` parallel methods (needs to be addressed before merge -- I'm working on it.)
- [ ] Rebase to current upstream (I will do this.)
- [ ] Identify source of the initial states discrepancy in the `compute_network_structured` and `compute_network_structured_obj` and correct (@hellkite500?)
- [ ] Port Reservoir code to parallel-friendly compute kernels -- at least `compute_network_structured`, and hopefully `compute_network` with the caching scheme.

File: src/python_framework_v02/troute/network/reach.pxd [hunk 3]
```diff
@@ -31,15 +31,20 @@ cdef struct _MC_Reach:
...
+cdef class MC_Reservoir(MC_Reach_Base_Class):
+  pass
+
+cdef class MC_Reach(MC_Reach_Base_Class):
   """
     A muskingcung reach -> collection of ordered MC_Segments
   """
   cdef _MC_Segment* _segments
   cdef int _num_segments # C only accessible
   cdef readonly num_segments #Python accessible, readonly outside class
```
Comment (jameshalgren) [inline src/python_framework_v02/troute/network/reach.pxd:48]: @groutr This is part of the structure...

**Comment by jameshalgren** [comment]
Following merge of PR #224, there will remain a few issues to manage in further development: 
- [ ] Resolve minor discrepancy due to handling of initial conditions
- [ ] Port Reservoir capability to the `compute_network_structured` function in mc_reach.pyx

**Review by hellkite500 — APPROVED** [review]

**Pull request merged**